### DISCUSSION TITLE
🚀 New Milestone in Posture Recognition Model's Performance.

### DISCUSSION THREAD

**Discussion by YanzhenHuang** [FYP-hyz-mjj-2024/posture_mmpose#21] [Show and tell]
## ✨ What we've done
- We applied per-channel normalization for data samples.
- We added a batch-normalization layer after all the convolution layers (before activation).
- We replaced the original MaxPool3D layer with our custom ResPool layer, which added the batch-normed data back to the original input  (so we named it the Residual Pooling layer). Unlike max-polling in 3D that loses 87.5% of the original information, ResPool only emphasized the local max from the original, boosting the model's learning performance.
- We removed the 4th convolution layer (C=32 -> C=32) and added a fully-connected layer. This prevents over-fitting.

## 👍 What's improved
- Training: 200 epoch - Min validation loss = 19.0367 => 25 epoch - Min validation loss = 11.0318
- Application: No more after-detection bias needed.

### Training Performance

<div style="display: flex; flex-wrap: wrap;">
<img src="https://github.com/user-attachments/assets/f69c45eb-c3e6-4676-b6a6-a4a112e220e7" height="250"/>
<img src="https://github.com/user-attachments/assets/998efb21-537b-43f0-b1c1-a7d8880427e7" height="250"/>
<img src="https://github.com/user-attachments/assets/d67606ed-a906-42bf-8fa8-418cd5cf5d1f" height="250"/>
<img src="https://github.com/user-attachments/assets/7bc5be23-a060-49bf-b011-d6e79efaa02d" height="250"/>
</div>

### Application Performance - Average computation time per frame (seconds)

<img src="https://github.com/user-attachments/assets/34e2e8b8-0726-4ebc-8872-d687931f5ba2" height="250"/>

**Comment by YanzhenHuang** [comment]
# Enhanced Weighted Loss introduced with MCLoss
## ✨ Key Points:
- To further improve the model's generalization ability, we applied the method of combining 1 main criterion and 2 associate criterions together as a WeightedLoss we used in classifier training.
	- The main criterion is CrossEntropyLoss.
	- The other 2 criterion is FocalLoss and L2_regularization.
	- The weights of criterions are 0.6, 0.3, and 0.1.

## 👀 Training Performance

Training early-stopped at:
```sh
Epoch[026/650], Train Loss:7.5009, Valid Loss:12.1534, OFF: 0.7858 | Cur Optim: 1967384384992, Min VL: 10.0571, Num OF epochs: 4
```

<div style="display: flex; flex-wrap: wrap;">
<img src="https://github.com/user-attachments/assets/be1cc700-0aa5-4675-a3fa-4ef7aff4a581" height="250"/>
<img src="https://github.com/user-attachments/assets/8e1abd5f-e2b8-408c-8478-508146f5ba91" height="250"/>
<img src="https://github.com/user-attachments/assets/40f7466d-91c0-40cf-92f4-c069762a7419" height="250"/>
<img src="https://github.com/user-attachments/assets/aa1c0a7f-a2d9-4a19-9b9f-2253b9ed3ed2" height="250"/>
</div>
